### PR TITLE
feat: support Trusted Types for client overlay

### PR DESCRIPTION
**Pull request by Siegrift**
- [ ] This is a **bugfix**
- [x] This is a **feature**
- [ ] This is a **code refactor**
- [ ] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?

Yes. LMK if more are needed.

### Motivation / Use-Case

Addresses https://github.com/webpack/webpack-dev-server/issues/4400

### Breaking Changes

No breaking changes. 

This change is noop for applications which do not enforce Trusted Types. Applications which do enforce Trusted Types and use webpack-dev-server can now whitelist `webpack-dev-server#overlay` policy to make error overlay work. Optionally, users can choose a different name via `policyName` attribute (but maybe this could change to more verbose `trustedTypesPolicyName`?). 

### Additional Info

See https://github.com/webpack/webpack-dev-server/issues/4400

### CODE REVIEW

File: client-src/index.js [hunk 1]
```diff
@@ -15,7 +15,7 @@ import createSocketURL from "./utils/createSocketURL.js";
  * @property {boolean} hot
  * @property {boolean} liveReload
  * @property {boolean} progress
- * @property {boolean | { warnings?: boolean, errors?: boolean }} overlay
+ * @property {boolean | { warnings?: boolean, errors?: boolean, policyName?: string }} overlay
```
Comment (Siegrift) [inline client-src/index.js:18]: I am wondering whether the `policyName` should be renamed to more verbose `trustedTypesPolicyName`. WDYT?

File: client-src/overlay.js [hunk 2]
```diff
@@ -23,10 +23,26 @@ let iframeContainerElement;
 let containerElement;
 /** @type {Array<(element: HTMLDivElement) => void>} */
 let onLoadQueue = [];
+/** @type {any} */
```
Comment (Siegrift) [inline client-src/overlay.js:26]: There are TS type definitions located in https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/trusted-types

However, I am not sure if it is worth typing so I've added `any` for now.

**Comment by alexander-akait** [comment]
Thanks for PR, can you accept CLA?

File: client-src/overlay.js [hunk 3]
```diff
@@ -23,10 +23,26 @@ let iframeContainerElement;
...
 
 ansiHTML.setColors(colors);
 
-function createContainer() {
+/**
+ * @param {string | null} policyName
+ */
+function createContainer(policyName) {
+  // Enable Trusted Types if they are available in the current browser.
+  if (window.trustedTypes) {
```
Comment (Siegrift) [inline client-src/overlay.js:37]: We could optionally check whether the `policyName` is not `null`. However, I think using trustedTypes whenever they are available is better because:

1. It is still noop and backwards compatible change for applications which do not enforce Trusted Types
2. This enables Trusted Types integration without the need to change anything in webpack configuration (which can be hidden, such as in [CRA](https://github.com/facebook/create-react-app)). This means that the only change the application needs to do is whitelist the default policy name.

File: types/lib/Server.d.ts [hunk 4]
```diff
@@ -1610,6 +1610,10 @@ declare class Server {
                     negatedDescription: string;
                   };
                 };
+                policyName: {
```
Comment (Siegrift) [inline types/lib/Server.d.ts:1613]: I am not sure why there are so many unrelated changes in this autogenerated `d.ts` file. Especially, since I've added just one field.

Please, LMK if I have something wrong in my setup.

**Comment by Siegrift** [comment]
I also see that there needs to be a documentation update in https://webpack.js.org/configuration/dev-server/#overlay, specifically this file https://github.com/1ilsang/webpack.js.org/blob/master/src/content/configuration/dev-server.mdx#overlay

I guess that is only updated after there is a new release?

Also, please LMK if there are is anything in this PR I've missed :)

File: client-src/index.js (hunk 1 above)
Comment (alexander-akait) [reply client-src/index.js:18]: I don't think we will have the `policyName` option here, but I think make sense to rename it to `trustedTypesPolicyName`, to avoid misleading

File: client-src/overlay.js (hunk 2 above)
Comment (alexander-akait) [reply client-src/overlay.js:26]: I think we need install them in dev depedencies, because when types is not found it is `any`

File: types/lib/Server.d.ts (hunk 4 above)
Comment (alexander-akait) [reply types/lib/Server.d.ts:1613]: It is bug in `ts`, ignore it

**Comment by alexander-akait** [comment]
> I also see that there needs to be a documentation update in https://webpack.js.org/configuration/dev-server/#overlay, specifically this file https://github.com/1ilsang/webpack.js.org/blob/master/src/content/configuration/dev-server.mdx#overlay

We will open issue in our docs after merge, dont worry

> I guess that is only updated after there is a new release?

Yes, I will do release

File: examples/client/trusted-types-overlay/README.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+```js
+module.exports = {
+  // ...
+  output: {
+    trustedTypes: { policyName: "webpack" },
+  },
+  devServer: {
+    client: {
+      overlay: {
+        policyName: "overlay-policy",
```
Comment (koto) [inline examples/client/trusted-types-overlay/README.md:14]: Consider using `webpack#dev-overlay` or something like this to feature the recommended naming scheme.

**Review by koto — DISMISSED** [review]

**Comment by alexander-akait** [comment]
@Siegrift frinedly ping

**Comment by Siegrift** [comment]
Sorry, I got a bit busy with work stuff this week. I'll hopefully find some time tomorrow or over the weekend. 

Thanks for the ping :)

**Comment by alexander-akait** [comment]
Bih thanks, feel free to ping me

File: client-src/index.js (hunk 1 above)
Comment (Siegrift) [reply client-src/index.js]: Fixed in https://github.com/webpack/webpack-dev-server/pull/4404/commits/c63489b3372539de2b51fffa738249435f5ff6bc

File: client-src/overlay.js (hunk 2 above)
Comment (Siegrift) [reply client-src/overlay.js]: Fixed https://github.com/webpack/webpack-dev-server/pull/4404/commits/68d39742f2c13dc3beb68913eecee0170f7a69af

File: examples/client/trusted-types-overlay/README.md (hunk 5 above)
Comment (Siegrift) [reply examples/client/trusted-types-overlay/README.md]: Fixed in https://github.com/webpack/webpack-dev-server/pull/4404/commits/8ef1a488fff474ff39097fc1394f6486b52b97b1

File: client-src/overlay.js [hunk 6]
```diff
@@ -23,10 +23,25 @@ let iframeContainerElement;
...
 ansiHTML.setColors(colors);
 
-function createContainer() {
+/**
+ * @param {string | null} trustedTypesPolicyName
+ */
+function createContainer(trustedTypesPolicyName) {
+  // Enable Trusted Types if they are available in the current browser.
+  if (window.trustedTypes) {
+    overlayTrustedTypesPolicy = window.trustedTypes.createPolicy(
```
Comment (Siegrift) [inline client-src/overlay.js:37]: In case there is an error creating a Trusted Types policy the overlay is not shown at all. 

A different approach would be to capture the exception outside this function (same place we define `overlayTrustedTypesPolicy`) and show an overlay without using `innerHTML` showing that the policy creation failed. 

However I think having an error in the console is sufficient, since this is a development only feature so we can assume the console is opened anyways.

**Comment by Siegrift** [comment]
I've fixed the comments and rebased on current `master` and left one more comment.

**Comment by alexander-akait** [comment]
@Siegrift Hello, sorry for delay, can you update snapshots?

**Comment by Siegrift** [comment]
Sorry for not running the tests locally first :/ 
However, there seem to be some changed snapshots which are unrelated to my feature. I am not sure about that.

**Comment by snitin315** [comment]
@Siegrift help output changed because you added `client-overlay-trusted-types-policy-name` CLI flag, hence need to update the snapshot.

**Comment by Siegrift** [comment]
Thanks, I am having problems passing the tests locally. When I run `yarn test:only --updateSnapshot` most of the tests pass, but there are sometimes a few unrelated changes (I've reverted those manually). When I try to run the tests now, they seem to remove the `basic should output help should generate correct cli flags 1` in webpack5 snapshot which is wrong.

Is there something I am doing wrong?

**Comment by snitin315** [comment]
Are you using macOS?

**Comment by Siegrift** [comment]
Yes, the m1 chip.

**Comment by snitin315** [comment]
Yeah, help output snapshots are not quite well on macOS, known bug in nodejs it strips the actual output on stdout.

**Comment by Siegrift** [comment]
I'll try to fix the tests on my Linux machine later today. I don't think I'll get anywhere by testing this on macOS :D

**Comment by alexander-akait** [comment]
I want to merge it and fix tests in the next PR, because we need to do release today due security problems in portfinder, big thanks for PR

**Comment by Siegrift** [comment]
No problem, sorry for taking it too long :/ 

Good luck with the release.
